neonvm/controllers/metrics.go

```diff
@@ -0,0 +1,71 @@
+package controllers
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type ReconcilerMetrics struct {
+	failing *prometheus.GaugeVec
+}
+
+func MakeReconcilerMetrics() ReconcilerMetrics {
+	m := ReconcilerMetrics{
+		failing: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "reconcile_failing_objects",
+				Help: "Number of objects that are failing to reconcile for each specific controller",
+			},
+			[]string{"controller"},
+		),
+	}
+	metrics.Registry.MustRegister(m.failing)
+	return m
+}
+
+type wrappedReconciler struct {
+	ControllerName string
+	Reconciler     reconcile.Reconciler
+	Metrics        ReconcilerMetrics
+
+	lock    sync.Mutex
+	failing map[client.ObjectKey]struct{}
+}
+
+// WithMetrics wraps a given Reconciler with metrics capabilities.
+func WithMetrics(reconciler reconcile.Reconciler, rm ReconcilerMetrics, cntrlName string) reconcile.Reconciler {
+	return &wrappedReconciler{
+		Reconciler:     reconciler,
+		Metrics:        rm,
+		ControllerName: cntrlName,
+		lock:           sync.Mutex{},
+		failing:        make(map[client.ObjectKey]struct{}),
+	}
+}
+
+func (d *wrappedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	res, err := d.Reconciler.Reconcile(ctx, req)
+
+	// This part is executed sequentially since we acquire a mutex lock. It
+	// should be quite fast since a mutex lock/unlock + 1 memory write takes
+	// ~60ns. Having 8 reconile workers, we wait at most 8*60=480ns to acquire
+	// the lock. I (@shayanh) preferred to go with the simplest implementation
+	// as of now. For a more performant solution, if needed, we can switch to an
+	// async approach.
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	if err != nil {
+		d.failing[req.NamespacedName] = struct{}{}
```

Comment:
Is it possible for an object to be deleted while still failing? (maybe with `kubectl delete --force`?)

If not, it'd be good to have a brief explainer why. If so, any ideas how we can avoid false positives in the failing count?

Reply:
I think if an object is deleted then it still goes through the `Reconcile` function but we treat it differently.

https://github.com/neondatabase/autoscaling/blob/ac66e2c7b11e7a8fd209f300a208de5d2003585a/neonvm/controllers/virtualmachine_controller.go#L140-L158

Reply:
Is that still guaranteed for `--force` ?

Reply:
Yes, `kubectl --force` doesn't wait for confirmation. It's a client-side behaviour. More info: https://kubernetes.io/docs/reference/kubectl/generated/kubectl_delete/

Reply:
Tried it out locally and I see that `--force` still respects the finalizer; fair enough